File: Dopamine.SettingsModule/ViewModels/SettingsPlaybackViewModel.vb
Imports System.Collections.ObjectModel
Imports Dopamine.Common.Services.Notification
Imports Dopamine.Common.Services.Playback
Imports Dopamine.Common.Services.Taskbar
Imports Dopamine.Core.Settings
Imports Dopamine.Core.Helpers
Imports Dopamine.Core.Utils
Imports Microsoft.Practices.Prism.Commands
Imports Microsoft.Practices.Prism.Mvvm
Imports Dopamine.Core.Base

Namespace ViewModels
    Public Class SettingsPlaybackViewModel
        Inherits BindableBase

#Region "Variables"
        Private mLatencies As ObservableCollection(Of NameValue)
        Private mSelectedLatency As NameValue
        Private mPlaybackService As IPlaybackService
        Private mTaskbarService As ITaskbarService
        Private mNotificationService As INotificationService
        'Private mCheckBoxWasapiEventModeChecked As Boolean
        Private mCheckBoxWasapiExclusiveModeChecked As Boolean
        Private mCheckBoxShowNotificationChecked As Boolean
        Private mCheckBoxShowNotificationControlsChecked As Boolean
        Private mCheckBoxShowProgressInTaskbarChecked As Boolean
        Private mCheckBoxShowNotificationOnlyWhenPlayerNotVisibleChecked As Boolean
        Private mNotificationPositions As ObservableCollection(Of NameValue)
        Private mSelectedNotificationPosition As NameValue
        Private mNotificationSeconds As ObservableCollection(Of Integer)
        Private mSelectedNotificationSecond As Integer
#End Region

#Region "Commands"
        Public Property ShowTestNotificationCommand As delegatecommand
#End Region

#Region "Properties"
        Public Property Latencies() As ObservableCollection(Of NameValue)
            Get
                Return mLatencies
            End Get
            Set(ByVal value As ObservableCollection(Of NameValue))
                SetProperty(Of ObservableCollection(Of NameValue))(mLatencies, value)
            End Set
        End Property

        Public Property SelectedLatency() As NameValue
            Get
                Return mSelectedLatency
            End Get
            Set(ByVal value As NameValue)
                XmlSettingsClient.Instance.Set(Of Integer)("Playback", "AudioLatency", value.Value)
                SetProperty(Of NameValue)(mSelectedLatency, value)

                If mPlaybackService IsNot Nothing Then
                    mPlaybackService.Latency = value.Value
                End If
            End Set
        End Property

        'Public Property CheckBoxWasapiEventModeChecked() As Boolean
        '    Get
        '        Return mCheckBoxWasapiEventModeChecked
        '    End Get
        '    Set(ByVal value As Boolean)
        '        XmlSettingsClient.Instance.Set(Of Boolean)("Playback", "WasapiEventMode", value)
        '        SetProperty(Of Boolean)(mCheckBoxWasapiEventModeChecked, value)

        '        If mPlaybackService IsNot Nothing Then
        '            mPlaybackService.EventMode = value
        '        End If
        '    End Set
        'End Property

        Public Property CheckBoxWasapiExclusiveModeChecked() As Boolean
            Get
                Return Me.mCheckBoxWasapiExclusiveModeChecked
            End Get
            Set(ByVal value As Boolean)
                XmlSettingsClient.Instance.Set(Of Boolean)("Playback", "WasapiExclusiveMode", value)
                SetProperty(Of Boolean)(mCheckBoxWasapiExclusiveModeChecked, value)

                If mPlaybackService IsNot Nothing Then
                    mPlaybackService.ExclusiveMode = value
                End If
            End Set
        End Property

        Public Property CheckBoxShowNotificationChecked() As Boolean
            Get
                Return mCheckBoxShowNotificationChecked
            End Get
            Set(ByVal value As Boolean)
                XmlSettingsClient.Instance.Set(Of Boolean)("Behaviour", "ShowNotification", value)
                SetProperty(Of Boolean)(mCheckBoxShowNotificationChecked, value)
            End Set
        End Property

        Public Property CheckBoxShowNotificationControlsChecked() As Boolean
            Get
                Return mCheckBoxShowNotificationControlsChecked
            End Get
            Set(ByVal value As Boolean)
                XmlSettingsClient.Instance.Set(Of Boolean)("Behaviour", "ShowNotificationControls", value)
                SetProperty(Of Boolean)(mCheckBoxShowNotificationControlsChecked, value)
            End Set
        End Property

        Public Property CheckBoxShowProgressInTaskbarChecked() As Boolean
            Get
                Return mCheckBoxShowProgressInTaskbarChecked
            End Get
            Set(ByVal value As Boolean)
                XmlSettingsClient.Instance.Set(Of Boolean)("Playback", "ShowProgressInTaskbar", value)
                SetProperty(Of Boolean)(Me.mCheckBoxShowProgressInTaskbarChecked, value)

                If mTaskbarService IsNot Nothing AndAlso mPlaybackService IsNot Nothing Then
                    mTaskbarService.SetTaskbarProgressState(value, mPlaybackService.IsPlaying)
                End If
            End Set
        End Property

        Public Property CheckBoxShowNotificationOnlyWhenPlayerNotVisibleChecked() As Boolean
            Get
                Return mCheckBoxShowNotificationOnlyWhenPlayerNotVisibleChecked
            End Get
            Set(ByVal value As Boolean)
                XmlSettingsClient.Instance.Set(Of Boolean)("Behaviour", "ShowNotificationOnlyWhenPlayerNotVisible", value)
                SetProperty(Of Boolean)(mCheckBoxShowNotificationOnlyWhenPlayerNotVisibleChecked, value)
            End Set
        End Property

        Public Property NotificationPositions() As ObservableCollection(Of NameValue)
            Get
                Return mNotificationPositions
            End Get
            Set(ByVal value As ObservableCollection(Of NameValue))
                SetProperty(Of ObservableCollection(Of NameValue))(mNotificationPositions, value)
            End Set
        End Property

        Public Property SelectedNotificationPosition() As NameValue
            Get
                Return mSelectedNotificationPosition
            End Get
            Set(ByVal value As NameValue)
                XmlSettingsClient.Instance.Set(Of Integer)("Behaviour", "NotificationPosition", value.Value)
                SetProperty(Of NameValue)(mSelectedNotificationPosition, value)
            End Set
        End Property

        Public Property NotificationSeconds() As ObservableCollection(Of Integer)
            Get
                Return mNotificationSeconds
            End Get
            Set(ByVal value As ObservableCollection(Of Integer))
                SetProperty(Of ObservableCollection(Of Integer))(mNotificationSeconds, value)
            End Set
        End Property

        Public Property SelectedNotificationSecond() As Integer
            Get
                Return mSelectedNotificationSecond
            End Get
            Set(ByVal value As Integer)
                XmlSettingsClient.Instance.Set(Of Integer)("Behaviour", "NotificationAutoCloseSeconds", value)
                SetProperty(Of Integer)(mSelectedNotificationSecond, value)
            End Set
        End Property
#End Region

#Region "Construction"
        Public Sub New(iPlaybackService As IPlaybackService, iTaskbarService As ITaskbarService, iNotificationService As INotificationService)

            mPlaybackService = iPlaybackService
            mTaskbarService = iTaskbarService
            mNotificationService = iNotificationService

            ShowTestNotificationCommand = New DelegateCommand(Sub() mNotificationService.ShowNotificationAsync())

            Me.GetCheckBoxesAsync()
            Me.GetNotificationPositionsAsync()
            Me.GetNotificationSecondsAsync()
            Me.GetLatenciesAsync()
        End Sub
#End Region

#Region "Private"
        Private Async Sub GetLatenciesAsync()

            Dim localLatencies As New ObservableCollection(Of NameValue)

            Await Task.Run(Sub()
                               For index = 50 To 500 Step 50 ' Increment by 50
                                   If index = 200 Then
                                       localLatencies.Add(New NameValue With {.Name = index & " ms (" & Application.Current.FindResource("Language_Default").ToString.ToLower & ")", .Value = index})
                                   Else
                                       localLatencies.Add(New NameValue With {.Name = index & " ms", .Value = index})
                                   End If
                               Next
                           End Sub)

            Me.Latencies = localLatencies

            Dim localSelectedLatency As NameValue = Nothing

            Await Task.Run(Sub() localSelectedLatency = Latencies.Where(Function(pa) pa.Value = XmlSettingsClient.Instance.Get(Of Integer)("Playback", "AudioLatency")).Select(Function(pa) pa).First())

            Me.SelectedLatency = localSelectedLatency
        End Sub

        Private Async Sub GetNotificationPositionsAsync()

            Dim localNotificationPositions As New ObservableCollection(Of NameValue)

            Await Task.Run(Sub()
                               localNotificationPositions.Add(New NameValue With {.Name = ResourceUtils.GetStringResource("Language_Bottom_Left"), .Value = NotificationPosition.BottomLeft})
                               localNotificationPositions.Add(New NameValue With {.Name = ResourceUtils.GetStringResource("Language_Top_Left"), .Value = NotificationPosition.TopLeft})
                               localNotificationPositions.Add(New NameValue With {.Name = ResourceUtils.GetStringResource("Language_Top_Right"), .Value = NotificationPosition.TopRight})
                               localNotificationPositions.Add(New NameValue With {.Name = ResourceUtils.GetStringResource("Language_Bottom_Right"), .Value = NotificationPosition.BottomRight})
                           End Sub)

            Me.NotificationPositions = localNotificationPositions

            Dim localSelectedNotificationPosition As NameValue = Nothing

            Await Task.Run(Sub() localSelectedNotificationPosition = NotificationPositions.Where(Function(np) np.Value = XmlSettingsClient.Instance.Get(Of Integer)("Behaviour", "NotificationPosition")).Select(Function(np) np).First())

            Me.SelectedNotificationPosition = localSelectedNotificationPosition
        End Sub

        Private Async Sub GetNotificationSecondsAsync()

            Dim localNotificationSeconds As New ObservableCollection(Of Integer)

            Await Task.Run(Sub()
                               For index = 1 To 5
                                   localNotificationSeconds.Add(index)
                               Next
                           End Sub)

            Me.NotificationSeconds = localNotificationSeconds

            Dim localSelectedNotificationSecond As Integer = 0

            Await Task.Run(Sub() localSelectedNotificationSecond = NotificationSeconds.Where(Function(ns) ns = XmlSettingsClient.Instance.Get(Of Integer)("Behaviour", "NotificationAutoCloseSeconds")).Select(Function(ns) ns).First())

            Me.SelectedNotificationSecond = localSelectedNotificationSecond
        End Sub

        Private Async Sub GetCheckBoxesAsync()

            Await Task.Run(Sub()
                               'Me.CheckBoxWasapiEventModeChecked = XmlSettingsClient.Instance.Get(Of Boolean)("Playback", "WasapiEventMode")
                               Me.CheckBoxWasapiExclusiveModeChecked = XmlSettingsClient.Instance.Get(Of Boolean)("Playback", "WasapiExclusiveMode")
                               Me.CheckBoxShowNotificationChecked = XmlSettingsClient.Instance.Get(Of Boolean)("Behaviour", "ShowNotification")
                               Me.CheckBoxShowNotificationControlsChecked = XmlSettingsClient.Instance.Get(Of Boolean)("Behaviour", "ShowNotificationControls")
                               Me.CheckBoxShowProgressInTaskbarChecked = XmlSettingsClient.Instance.Get(Of Boolean)("Playback", "ShowProgressInTaskbar")
                               Me.CheckBoxShowNotificationOnlyWhenPlayerNotVisibleChecked = XmlSettingsClient.Instance.Get(Of Boolean)("Behaviour", "ShowNotificationOnlyWhenPlayerNotVisible")
                           End Sub)

        End Sub
#End Region
    End Class
End Namespace

File: Dopamine.SettingsModule/ViewModels/SettingsPlaybackViewModel.cs
using System;
using Digimezzo.Utilities.Helpers;
using Digimezzo.Utilities.Settings;
using Digimezzo.Utilities.Utils;
using Dopamine.Common.Services.Dialog;
using Dopamine.Common.Services.Notification;
using Dopamine.Common.Services.Playback;
using Dopamine.Common.Services.Taskbar;
using Dopamine.Common.Base;
using Prism.Commands;
using Prism.Mvvm;
using System.Collections.ObjectModel;
using System.Linq;
using System.Threading.Tasks;
using System.Windows;
using CSCore.CoreAudioAPI;
using System.Collections.Generic;
using Dopamine.Common.Services.ExternalControl;

namespace Dopamine.SettingsModule.ViewModels
{
    public class SettingsPlaybackViewModel : BindableBase
    {
        #region DeviceContainer
        public class OutputDevice
        {
            public string Name { get; set; }
            public MMDevice Device { get; set; }

            public override string ToString()
            {
                return this.Name;
            }

            public override bool Equals(object obj)
            {
                if (obj == null || !GetType().Equals(obj.GetType()))
                {
                    return false;
                }

                return this.Name.Equals(((OutputDevice)obj).Name);
            }

            public override int GetHashCode()
            {
                return new { this.Name, }.GetHashCode();
            }
        }
        #endregion

        #region Variables
        private ObservableCollection<NameValue> latencies;
        private NameValue selectedLatency;
        private IPlaybackService playbackService;
        private ITaskbarService taskbarService;
        private INotificationService notificationService;
        private IDialogService dialogService;
        protected IExternalControlService externalControlService;
        private bool checkBoxWasapiExclusiveModeChecked;
        private bool checkBoxShowNotificationWhenPlayingChecked;
        private bool checkBoxShowNotificationWhenPausingChecked;
        private bool checkBoxShowNotificationWhenResumingChecked;
        private bool checkBoxShowNotificationControlsChecked;
        private bool checkBoxShowProgressInTaskbarChecked;
        private bool checkBoxShowNotificationOnlyWhenPlayerNotVisibleChecked;
        private bool checkBoxEnableExternalControlChecked;
        private bool checkBoxEnableSystemNotificationChecked;
        private ObservableCollection<NameValue> notificationPositions;
        private NameValue selectedNotificationPosition;
        private ObservableCollection<int> notificationSeconds;
        private int selectedNotificationSecond;
        private ObservableCollection<OutputDevice> outputDevices;
        private OutputDevice selectedOutputDevice;
        #endregion

        #region Commands
        public DelegateCommand ShowTestNotificationCommand { get; set; }
        #endregion

        #region Properties
        public bool IsNotificationEnabled => this.CheckBoxShowNotificationWhenPlayingChecked || this.CheckBoxShowNotificationWhenPausingChecked || this.CheckBoxShowNotificationWhenResumingChecked;

        public ObservableCollection<NameValue> Latencies
        {
            get => this.latencies;
            set => SetProperty<ObservableCollection<NameValue>>(ref this.latencies, value);
        }

        public NameValue SelectedLatency
        {
            get => this.selectedLatency;
            set
            {
                SettingsClient.Set<int>("Playback", "AudioLatency", value.Value);
                SetProperty<NameValue>(ref this.selectedLatency, value);

                if (this.playbackService != null)
                {
                    this.playbackService.Latency = value.Value;
                }
            }
        }

        public bool CheckBoxWasapiExclusiveModeChecked
        {
            get { return this.checkBoxWasapiExclusiveModeChecked; }
            set
            {
                if (value)
                {
                    this.ConfirmEnableExclusiveMode();
                }
                else
                {
                    this.ApplyExclusiveMode(false);
                }
            }
        }

        public bool CheckBoxShowNotificationWhenPlayingChecked
        {
            get => this.checkBoxShowNotificationWhenPlayingChecked;
            set
            {
                SettingsClient.Set<bool>("Behaviour", "ShowNotificationWhenPlaying", value);
                SetProperty<bool>(ref this.checkBoxShowNotificationWhenPlayingChecked, value);
                RaisePropertyChanged(nameof(this.IsNotificationEnabled));
            }
        }

        public bool CheckBoxShowNotificationWhenPausingChecked
        {
            get => this.checkBoxShowNotificationWhenPausingChecked;
            set
            {
                SettingsClient.Set<bool>("Behaviour", "ShowNotificationWhenPausing", value);
                SetProperty<bool>(ref this.checkBoxShowNotificationWhenPausingChecked, value);
                RaisePropertyChanged(nameof(this.IsNotificationEnabled));
            }
        }

        public bool CheckBoxShowNotificationWhenResumingChecked
        {
            get => this.checkBoxShowNotificationWhenResumingChecked;
            set
            {
                SettingsClient.Set<bool>("Behaviour", "ShowNotificationWhenResuming", value);
                SetProperty<bool>(ref this.checkBoxShowNotificationWhenResumingChecked, value);
                RaisePropertyChanged(nameof(this.IsNotificationEnabled));
            }
        }

        public bool CheckBoxShowNotificationControlsChecked
        {
            get => this.checkBoxShowNotificationControlsChecked;
            set
            {
                SettingsClient.Set<bool>("Behaviour", "ShowNotificationControls", value);
                SetProperty<bool>(ref this.checkBoxShowNotificationControlsChecked, value);
            }
        }

        public bool CheckBoxShowProgressInTaskbarChecked
        {
            get => this.checkBoxShowProgressInTaskbarChecked;
            set
            {
                SettingsClient.Set<bool>("Playback", "ShowProgressInTaskbar", value);
                SetProperty<bool>(ref this.checkBoxShowProgressInTaskbarChecked, value);

                if (this.taskbarService != null && this.playbackService != null)
                {
                    this.taskbarService.SetShowProgressInTaskbar(value);
                }
            }
        }

        public bool CheckBoxShowNotificationOnlyWhenPlayerNotVisibleChecked
        {
            get => this.checkBoxShowNotificationOnlyWhenPlayerNotVisibleChecked;
            set
            {
                SettingsClient.Set<bool>("Behaviour", "ShowNotificationOnlyWhenPlayerNotVisible", value);
                SetProperty<bool>(ref this.checkBoxShowNotificationOnlyWhenPlayerNotVisibleChecked, value);
            }
        }

        public ObservableCollection<NameValue> NotificationPositions
        {
            get => this.notificationPositions;
            set => SetProperty<ObservableCollection<NameValue>>(ref this.notificationPositions, value);
        }

        public NameValue SelectedNotificationPosition
        {
            get => this.selectedNotificationPosition;
            set
            {
                SettingsClient.Set<int>("Behaviour", "NotificationPosition", value.Value);
                SetProperty<NameValue>(ref this.selectedNotificationPosition, value);
            }
        }

        public ObservableCollection<int> NotificationSeconds
        {
            get => this.notificationSeconds;
            set => SetProperty<ObservableCollection<int>>(ref this.notificationSeconds, value);
        }

        public int SelectedNotificationSecond
        {
            get => this.selectedNotificationSecond;
            set
            {
                SettingsClient.Set<int>("Behaviour", "NotificationAutoCloseSeconds", value);
                SetProperty<int>(ref this.selectedNotificationSecond, value);
            }
        }

        public ObservableCollection<OutputDevice> OutputDevices
        {
            get => this.outputDevices;
            set => SetProperty<ObservableCollection<OutputDevice>>(ref this.outputDevices, value);
        }

        public OutputDevice SelectedOutputDevice
        {
            get => this.selectedOutputDevice;
            set
            {
                SetProperty<OutputDevice>(ref this.selectedOutputDevice, value);

                // Due to two-way binding, this can be null when the list is being filled.
                if (value != null)
                {
                    SettingsClient.Set<string>("Playback", "AudioDevice", value.Device == null ? string.Empty : value.Device.DeviceID);

                    this.playbackService.SwitchOutputDeviceAsync(value.Device);
                } 
            }
        }

        public bool CheckBoxEnableExternalControlChecked
        {
            get => this.checkBoxEnableExternalControlChecked;
            set
            {
                SettingsClient.Set("Playback", "EnableExternalControl", value);
                SetProperty(ref this.checkBoxEnableExternalControlChecked, value);
                if (value == true)
                    this.externalControlService.Start();
                else
                    this.externalControlService.Stop();
            }
        }

        public bool IsWindows10 => EnvironmentUtils.IsWindows10();

        public bool CheckBoxEnableSystemNotificationChecked
        {
            get => this.checkBoxEnableSystemNotificationChecked;
            set
            {
                SettingsClient.Set("Behaviour", "EnableSystemNotification", value);
                SetProperty(ref this.checkBoxEnableSystemNotificationChecked, value);
            }
        }

        #endregion

        #region Construction
        public SettingsPlaybackViewModel(IPlaybackService playbackService, ITaskbarService taskbarService, INotificationService notificationService, IDialogService dialogService, IExternalControlService externalControlService)
        {
            this.playbackService = playbackService;
            this.taskbarService = taskbarService;
            this.notificationService = notificationService;
            this.dialogService = dialogService;
            this.externalControlService = externalControlService;

            ShowTestNotificationCommand = new DelegateCommand(() => this.notificationService.ShowNotificationAsync());

            this.playbackService.AudioDevicesChanged += (_,__) => Application.Current.Dispatcher.Invoke(() => this.GetOutputDevicesAsync());

            this.GetCheckBoxesAsync();
            this.GetNotificationPositionsAsync();
            this.GetNotificationSecondsAsync();
            this.GetLatenciesAsync();
            this.GetOutputDevicesAsync();
        }
        #endregion

        #region Private
        private async void GetOutputDevicesAsync()
        {
            IList<MMDevice> outputDevices = await this.playbackService.GetAllOutputDevicesAsync();

            this.OutputDevices = new ObservableCollection<OutputDevice>();

            this.OutputDevices.Add(new OutputDevice() { Name = ResourceUtils.GetStringResource("Language_Default_Audio_Device"), Device = null });

            foreach (MMDevice device in outputDevices)
            {
                this.OutputDevices.Add(new OutputDevice() { Name = device.FriendlyName, Device = device });
            }

            MMDevice savedDevice = await this.playbackService.GetSavedAudioDeviceAsync();

            this.selectedOutputDevice = savedDevice == null ? this.OutputDevices.First() : new OutputDevice() { Name = savedDevice.FriendlyName, Device = savedDevice };
            RaisePropertyChanged(nameof(this.SelectedOutputDevice));
        }

        private async void GetLatenciesAsync()
        {
            var localLatencies = new ObservableCollection<NameValue>();

            await Task.Run(() =>
            {
                // Increment by 50
                for (int index = 50; index <= 500; index += 50)
                {
                    if (index == 200)
                    {
                        localLatencies.Add(new NameValue
                        {
                            Name = index + " ms (" + Application.Current.FindResource("Language_Default").ToString().ToLower() + ")",
                            Value = index
                        });
                    }
                    else
                    {
                        localLatencies.Add(new NameValue
                        {
                            Name = index + " ms",
                            Value = index
                        });
                    }
                }
            });

            this.Latencies = localLatencies;

            NameValue localSelectedLatency = null;

            await Task.Run(() => localSelectedLatency = this.Latencies.Where((pa) => pa.Value == SettingsClient.Get<int>("Playback", "AudioLatency")).Select((pa) => pa).First());

            this.SelectedLatency = localSelectedLatency;
        }

        private async void GetNotificationPositionsAsync()
        {
            var localNotificationPositions = new ObservableCollection<NameValue>();

            await Task.Run(() =>
            {
                localNotificationPositions.Add(new NameValue { Name = ResourceUtils.GetStringResource("Language_Bottom_Left"), Value = (int)NotificationPosition.BottomLeft });
                localNotificationPositions.Add(new NameValue { Name = ResourceUtils.GetStringResource("Language_Top_Left"), Value = (int)NotificationPosition.TopLeft });
                localNotificationPositions.Add(new NameValue { Name = ResourceUtils.GetStringResource("Language_Top_Right"), Value = (int)NotificationPosition.TopRight });
                localNotificationPositions.Add(new NameValue { Name = ResourceUtils.GetStringResource("Language_Bottom_Right"), Value = (int)NotificationPosition.BottomRight });
            });

            this.NotificationPositions = localNotificationPositions;

            NameValue localSelectedNotificationPosition = null;

            await Task.Run(() => localSelectedNotificationPosition = NotificationPositions.Where((np) => np.Value == SettingsClient.Get<int>("Behaviour", "NotificationPosition")).Select((np) => np).First());

            this.SelectedNotificationPosition = localSelectedNotificationPosition;
        }

        private async void GetNotificationSecondsAsync()
        {
            var localNotificationSeconds = new ObservableCollection<int>();

            await Task.Run(() =>
            {
                for (int index = 1; index <= 5; index++)
                {
                    localNotificationSeconds.Add(index);
                }

            });

            this.NotificationSeconds = localNotificationSeconds;

            int localSelectedNotificationSecond = 0;

            await Task.Run(() => localSelectedNotificationSecond = NotificationSeconds.Where((ns) => ns == SettingsClient.Get<int>("Behaviour", "NotificationAutoCloseSeconds")).Select((ns) => ns).First());

            this.SelectedNotificationSecond = localSelectedNotificationSecond;
        }

        private async void GetCheckBoxesAsync()
        {
            await Task.Run(() =>
            {
                // Change the backing field, not the property. Otherwise the confirmation popup is shown when the screen is constructed.
                this.checkBoxWasapiExclusiveModeChecked = SettingsClient.Get<bool>("Playback", "WasapiExclusiveMode");
                RaisePropertyChanged(nameof(this.CheckBoxWasapiExclusiveModeChecked));

                this.CheckBoxShowNotificationWhenPlayingChecked = SettingsClient.Get<bool>("Behaviour", "ShowNotificationWhenPlaying");
                this.CheckBoxShowNotificationWhenPausingChecked = SettingsClient.Get<bool>("Behaviour", "ShowNotificationWhenPausing");
                this.CheckBoxShowNotificationWhenResumingChecked = SettingsClient.Get<bool>("Behaviour", "ShowNotificationWhenResuming");
                this.CheckBoxShowNotificationControlsChecked = SettingsClient.Get<bool>("Behaviour", "ShowNotificationControls");
                this.CheckBoxShowProgressInTaskbarChecked = SettingsClient.Get<bool>("Playback", "ShowProgressInTaskbar");
                this.CheckBoxShowNotificationOnlyWhenPlayerNotVisibleChecked = SettingsClient.Get<bool>("Behaviour", "ShowNotificationOnlyWhenPlayerNotVisible");
                this.checkBoxEnableExternalControlChecked = SettingsClient.Get<bool>("Playback", "EnableExternalControl");
            });
        }

        private void ConfirmEnableExclusiveMode()
        {
            if (this.dialogService.ShowConfirmation(
                0xe11b,
                16,
                ResourceUtils.GetStringResource("Language_Exclusive_Mode"),
                ResourceUtils.GetStringResource("Language_Exclusive_Mode_Confirmation"),
                ResourceUtils.GetStringResource("Language_Yes"),
                ResourceUtils.GetStringResource("Language_No")))
            {
                ApplyExclusiveMode(true);
            }
        }

        private void ApplyExclusiveMode(bool isEnabled)
        {
            SettingsClient.Set<bool>("Playback", "WasapiExclusiveMode", isEnabled);
            SetProperty<bool>(ref this.checkBoxWasapiExclusiveModeChecked, isEnabled);
            if (this.playbackService != null) this.playbackService.ExclusiveMode = isEnabled;
        }
        #endregion
    }
}
